### PR TITLE
Grammar: adds support for CREATE/ALTER/DROP DATABASE for Postgres dialect

### LINKED ISSUE #2017: Alter Database Unparsable Section
## Expected Behaviour
Using the postgres [ALTER DATABASE](https://www.postgresql.org/docs/14/sql-alterdatabase.html) should work.

## Observed Behaviour
```bash
$ sqlfluff lint --dialect postgres alter.sql
== [alter.sql] FAIL                                                          
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'ALTER
                       | DATABASE postgres SET password_enc...'
All Finished 📜 🎉!
```

## Steps to Reproduce

**alter.sql**:

```sql
ALTER DATABASE postgres SET password_encryption TO 'scram-sha-256';
```

```bash
sqlfluff lint --dialect postgres alter.sql
```

## Dialect
`postgres`

## Version
`sqlfluff, version 0.8.2`
`Python 3.8.10`
`pip 20.0.2`

## Configuration
`--dialtect postgres`

### PR DESCRIPTION
**Pull request by derickl**
### Brief summary of the change made
Adds full support for [CREATE | ALTER | DROP] DATABASE statements.

Also fixes #2017 

### Are there any other side effects of this change that we should be aware of?
None

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Review by jpy-git — COMMENTED** [review]
@derickl a couple things for CREATE/DROP whilst I'm still reviewing ALTER 👍

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 1]
```diff
@@ -1555,6 +1555,159 @@ class DropMaterializedViewStatementSegment(BaseSegment):
...
+                Sequence("WITH", optional=True),
+                AnyNumberOf(
+                    Sequence("ALLOW_CONNECTIONS", Ref("BooleanLiteralGrammar")),
+                    Sequence(
+                        "CONNECTION",
+                        "LIMIT",
+                        Ref("NumericLiteralSegment"),
+                    ),
+                    Sequence("IS_TEMPLATE", Ref("BooleanLiteralGrammar")),
+                ),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1660]:
```suggestion
                AnyNumberOf(
                    Sequence("ALLOW_CONNECTIONS", Ref("BooleanLiteralGrammar")),
                    Sequence(
                        "CONNECTION",
                        "LIMIT",
                        Ref("NumericLiteralSegment"),
                    ),
                    Sequence("IS_TEMPLATE", Ref("BooleanLiteralGrammar")),
                    min_times=1,
                ),
```
slight nuance here, we can have any number of these but we must have at least one of them so we use the min_times argument to AnyNumberOf

**Review by jpy-git — COMMENTED** [review]
A couple more suggestions for ALTER DATABASE.

It's looking great so far! 🙌

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 2]
```diff
@@ -1555,6 +1555,159 @@ class DropMaterializedViewStatementSegment(BaseSegment):
...
+
+    type = "create_database_statement"
+
+    match_grammar = StartsWith(Sequence("CREATE", "DATABASE"))
+
+    parse_grammar = Sequence(
+        "CREATE",
+        "DATABASE",
+        Ref("DatabaseReferenceSegment"),
+        Sequence("WITH", optional=True),
```
[the remark this answers is not in the document]
Comment (derickl) [reply src/sqlfluff/dialects/dialect_postgres.py:1573]: I've grep'd through the source code and haven't come across this usage pattern `Ref("WITH", optional=True)` but `Ref.keyword("WITH", optional=True)` does exist.

When does one choose `Ref.keyword` over `Sequence`? Would `Ref.keyword` be for `reserved` keywords?

**Comment by derickl** [comment]
Based on this https://github.com/sqlfluff/sqlfluff/pull/2081#discussion_r766207381

I've grep'd through the source code and haven't come across this usage pattern `Ref("WITH", optional=True)` but `Ref.keyword("WITH", optional=True)` does exist. 

What would be the difference between `Ref("WITH", optional=True)` and `Sequence("WITH", optional=True)`?

When does one choose `Ref.keyword` over `Sequence`? Would `Ref.keyword` be for `reserved` keywords?

**Comment by derickl** [comment]
Thank you for the review @jpy-git!

**Comment by tunetheweb** [comment]
> Based on this [#2081 (comment)](https://github.com/sqlfluff/sqlfluff/pull/2081#discussion_r766207381)
> 
> I've grep'd through the source code and haven't come across this usage pattern `Ref("WITH", optional=True)` but `Ref.keyword("WITH", optional=True)` does exist.
> 
> What would be the difference between `Ref("WITH", optional=True)` and `Sequence("WITH", optional=True)`?
> 
> When does one choose `Ref.keyword` over `Sequence`? Would `Ref.keyword` be for `reserved` keywords?

A good question and one I’ve wondered in the past!

As far as I can tell there is no real difference and it comes down to personal preference.

In theory using `Ref.keyword` might be slightly faster as the Sequence version effectively wraps the implied `Ref.keyword` in a (short) `Sequence` So it’s an extra layer to go through. In practice I would be VERY surprised if that created any real difference.

Personally I have a slight preference for Sequence (despite the above) for a few reasons:
- it seems more recognisable as we use Sequenxe a lot.
- It allows other grammar to more easily be added to this section (e.g. if you want to add a `NOT` before it).
- It two less characters and all one word so easier to type!

But honestly it doesn’t make a big difference. Maybe we should standardise on a preferred style but when I looked before they were both used quite a bit and I’ll bet it’s got worse since then.

File: test/fixtures/dialects/postgres/postgres_create_database.sql [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+CREATE DATABASE db OWNER = user_name;
+CREATE DATABASE db WITH OWNER = user_name;
+CREATE DATABASE db TEMPLATE template_name;
+CREATE DATABASE db TEMPLATE = template_name;
+CREATE DATABASE db WITH TEMPLATE = template_name;
+CREATE DATABASE db ENCODING 'UTF8';
+CREATE DATABASE db ENCODING = 'UTF8';
+CREATE DATABASE db LOCALE 'en_US.UTF-8';
+CREATE DATABASE db LOCALE = 'en_US.utf8';
+CREATE DATABASE db LC_COLLATE 'en_US.UTF-8' LC_CTYPE 'en_US.UTF-8';
```
Comment (jpy-git) [inline test/fixtures/dialects/postgres/postgres_create_database.sql:12]: could we have a unit test with the order of `LC_COLLATE` and `LC_CTYPE` switched to cover the point we discussed in the previous review?

File: test/fixtures/dialects/postgres/postgres_create_database.sql [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+CREATE DATABASE db TABLESPACE = new_tablespace;
+CREATE DATABASE db WITH TABLESPACE = new_tablespace;
+CREATE DATABASE db ALLOW_CONNECTIONS true;
+CREATE DATABASE db ALLOW_CONNECTIONS = true;
+CREATE DATABASE db CONNECTION LIMIT 10;
+CREATE DATABASE db CONNECTION LIMIT = 10;
+CREATE DATABASE db IS_TEMPLATE true;
+CREATE DATABASE db IS_TEMPLATE = true;
+CREATE DATABASE db IS_TEMPLATE = true CONNECTION LIMIT 10 ALLOW_CONNECTIONS = true;
+CREATE DATABASE db WITH IS_TEMPLATE = true CONNECTION LIMIT 10 ALLOW_CONNECTIONS = true;
```
Comment (jpy-git) [inline test/fixtures/dialects/postgres/postgres_create_database.sql:28]: Wonder if we should have one test case with all of the options specified to really test the grammar?
It should work but good to check 😄

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 5]
```diff
@@ -1555,6 +1555,166 @@ class DropMaterializedViewStatementSegment(BaseSegment):
...
+
+    match_grammar = StartsWith(Sequence("ALTER", "DATABASE"))
+
+    parse_grammar = Sequence(
+        "ALTER",
+        "DATABASE",
+        Ref("DatabaseReferenceSegment"),
+        OneOf(
+            Sequence(
+                Sequence("WITH", optional=True),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1653]:
```suggestion
                Ref.keyword("WITH", optional=True),
```
To be consistent with CREATE/DROP

File: test/fixtures/dialects/postgres/postgres_alter_database.sql [hunk 6]
```diff
@@ -0,0 +1,29 @@
+ALTER DATABASE db;
+ALTER DATABASE db ALLOW_CONNECTIONS true;
+ALTER DATABASE db WITH ALLOW_CONNECTIONS true;
+ALTER DATABASE db CONNECTION LIMIT 10;
+ALTER DATABASE db WITH CONNECTION LIMIT 10;
+ALTER DATABASE db IS_TEMPLATE true;
+ALTER DATABASE db WITH IS_TEMPLATE true;
+ALTER DATABASE db IS_TEMPLATE true ALLOW_CONNECTIONS true;
+ALTER DATABASE db WITH IS_TEMPLATE true ALLOW_CONNECTIONS true;
```
Comment (jpy-git) [inline test/fixtures/dialects/postgres/postgres_alter_database.sql:9]: Maybe a test case with ALLOW_CONNECTIONS, CONNECTION LIMIT, and IS_TEMPLATE all specified together?

**Review by jpy-git — COMMENTED** [review]
Looks good, couple final things to add then we should be good to get this going 👍

**Comment by jpy-git** [comment]
@derickl good spot on the `Ref` vs `Ref.keyword`! 

Basically Ref.keyword is the syntax for explicitly specifying to search for the string in the keywords. You can actually think of any of the keywords as being wrapped by Ref.keyword
e.g.
```sql
parse_grammar = Sequence(
    Ref.keyword("CREATE"),
    Ref.keyword("DATABASE"),
...
```
but for shorthand you can exclude `Ref.keyword` in these cases. 

However, in the specific example of the WITH in this PR we need a way to tell the parser that it is optional. That's why we pass the keyword via the more explicit `Ref.keyword` segment so that we can also pass an `optional=True` argument.

I agree that Sequence could work but think Ref.keyword is more accurate to the logic we're trying to express (the maintainers can have a discussion around consistency outside of this PR).

File: test/fixtures/dialects/postgres/postgres_create_database.sql (hunk 3 above)
Comment (derickl) [reply test/fixtures/dialects/postgres/postgres_create_database.sql:12]: I missed that!

File: test/fixtures/dialects/postgres/postgres_create_database.sql (hunk 4 above)
Comment (derickl) [reply test/fixtures/dialects/postgres/postgres_create_database.sql:28]: Will do

**Review by jpy-git — APPROVED** [review]
Awesome work @derickl and really robust test cases, LGTM 🚀 🌟

**Comment by derickl** [comment]
Thanks for the great review @jpy-git

**Comment by derickl** [comment]
Thank you for sharing your thoughts [here](https://github.com/sqlfluff/sqlfluff/pull/2081#issuecomment-990708401)  @tunetheweb. I hope we converge to some consistency down the road.

**Comment by tunetheweb** [comment]
> I hope we converge to some consistency down the road.

Or maybe not if @jpy-git and I can't even agree! 🤣

**Pull request merged**